### ISSUE TITLE
Common handling of authentication 

### ISSUE THREAD

**Issue by lipkau** [AtlassianPS/AtlassianPS#4]
All AtlassianPS modules should have the same handling for the authentication:
* all functions that connect to the server should have a `-ApiURi` parameter
* all functions that connect to the server should have a `-Credential` parameter
* all modules should have a `New-Session` command (that uses basic authentication and stores the session variable)
* all functions that connect to the server should allow for the server and credentials to be provided by a config file
* all functions that connect to the server should have a parameter to choose from the server in the confit file by a friendly name

**Comment by lipkau** [comment]
One PowerShell session can have multiple Jira/Confluence Servers at a given time and a default connection can be set

```powershell
Set-Configuration -Uri 'http://powershell.atlassian.net' -Alias 'me@cloud' -Credential $user
Set-Configuration -Uri 'http://powershell.atlassian.net' -Alias 'admin@cloud' -Credential $admin
Connect-Server -Server 'me@cloud'
Get-SOMETHING                             # uses 'me@cloud'
Get-SOMETHING -Server 'admin@cloud'       # uses 'admin@cloud'
```
> This let's you use *Option 1* but still gives you the option to _additionally_ use *Option 2* to set the `$PSDefaultParameterValues`.
> The downside to *Option 2*, is that it would require you to configure the server and connect -> 2 commands ... compared to the 1 command in *Option 2*
> This is aligned with the current implementation in `ConfluencePS` but is a breaking change for `JiraPS`.

**Comment by lipkau** [comment]
## Preamble

Currently all AtlassianPS modules have a different way of handling how they connect to the server and store credentials

* JiraPS created a `config.xml` file in the root of the module and is able to store a session
* ConfluencePS accepts `URI` and `Credentials` on all functions and has a function to set `$PSDefaultParameterValues` for them
* BitbucketPS has a first implementation of what I am about to propose

## Proposition

### TL;DR

Have a new AtlassianPS module (whish does nothing on it's own) which will handle all the servers and credentials.

Let's call it `AtlassianPS.Configuration` for now

### The basics

Each module would require `AtlassianPS.Configuration` in it's manifest.

The module will have proxy functions to `AtlassianPS.Configuration`'s functions which _can_ add a filter.

`AtlassianPS.Configuration` will be responsible to store and manage the data.

### Architecture of the Configuration

`AtlassianPS.Configuration` will use `PoshCode\Configuration` or `PowershellFrameworkCollective/PSFramework` to import and export the data (dependency declared in the manifest).

When loading the module using `AtlassianPS.Configuration`, `AtlassianPS.Configuration` will import any stored configuration.

The module using `AtlassianPS.Configuration` will use `AtlassianPS.Configuration\Get-Configuration` when it requires a value stored in the configuration.

### Structure of the Configuration

Stored in `AtlassianPS.Configuration`

```powershell
$script:Configuration = @{
    "Server" = @(
        [AtlassianPS.ServerConfiguration]@{
            "Name": "Server Prod",
            "Uri": [Uri]"https://server.com/",
            "Session": null,
            "Type": [AtlassianPS.ServerType]"jira" | "confluence" | "bitbucket",
            "IsCloudServer": true | false,
            "Header": {}
        }
    ),
    "Headers" = @{}
}
```

### AtlassianPS.Configuration Cmdlets

* `Export-Configuration`: saves the current config to disc
* `Get-ServerConfiguration`: retrieves a specific `[AtlassianPS.ServerConfiguration]` from the config
* `Set-ServerConfiguration`: stores a `[AtlassianPS.ServerConfiguration]` to the config (overwrites existing)
* `Remove-ServerConfiguration`: removes a specific `[AtlassianPS.ServerConfiguration]`
* `Get-Configuration`: retrieves a specific key in the config
* `Set-Configuration`: sets a new value for a key in the config
* `Remove-Configuration`: removes a key from the config

**Comment by dezren39** [comment]
I've been content with the JiraPS auth system so far, but this seems like a great abstraction for this set of modules, even if I have to redo stuff that won't improve my current situation. I may just wait a bit before updating again if it's a bunch of refactoring.

I am kinda wondering if the Cmdlets should denote Atlassian? Export-AtlassianConfiguration or something? In PowerShell that seems common, especially with many modules sharing one PowerShell. In Java/Clean Code, the format above without Atlassian would make a lot more sense and would be my preference.

**Comment by lipkau** [comment]
This is a breaking change, and will require changes in all script using it.

Fortunately it's _only_ the part defining the connection. 

---

The sample `Export-Configuration` is missing it's prefix.

I should have mentioned that... the module by itself would be `Export-AtlassianConfiguration`.

But as the main intention is to run it with the normal module (eg JiraPS), the command would be `Export-JiraConfiguration`

**Comment by ritzcrackr** [comment]
@lipkau,  I was getting ready to submit this proposal so I could get you feedback and I found this issue. I really your proposal above as it very similar to what I am proposing below. Is there a reason you didn't go forward with your proposal? As I mention below, the v3 effort would be a perfect time for big changes. I don't think adding the parameters below would be a break change but I might have missed something. 

Ultimately my proposal is about ensuring whatever changes I try to make are consistent with how you see the project going forward. Please let me know how you would like this project to proceed. Thanks.

# Proposal for Handling Multiple Authentication, Sessions and Common Parameters

In order to provide a roadmap for changes and ensure that work being done will be accepted by this project I am making the following proposal.

The idea behind this proposal is to provide a unified concept around the common parameters that all publicly exposed Cmdlets must have and the Session Cmdlets in the AtlassianPS.Configuration.
These parameters include both authentication and the Rest API endpoint that the cmdlets will operate against. The goal being to simplify and concentrate the handling of Authentication, Sessions, and Common Parameters while keeping each cmdlet focused on its function.
The ideas for this were based on the patterns found in the *-PSSession cmdlets and the *-WebRequest/*-RestMethod Cmdlets.
For example, the New-PSSession cmdlet will create a session object with the authentication and target machine that the rest of the *-PSSession cmdlets can use. At the time both Connect-PSSession and Enter-PSSession can leverage the session object created by New-PSSession or a user can use the authentication and destination parameters each of these cmdlets also expose.

Below I have defined the common parameters that each public cmdlet should expose. I have defined some suggested helper functions that make handling these parameters seamless and consistent for each cmdlet.

Definitions:
Sessions - Configuration/Sessions. The AtlassianPS.Configuration Module seems to be all about gettinng, setting, exporting and importing configuration/sessions for Atlassian 
Servers. It seems like the *-PSSession Cmdlets would be a good example of how this should work.

## Common Parameters

* Exposed by all Public Cmdlets, i.e. Get-ConfluencePage or Get-JiraIssue

### -Credential

* Type
  * [System.Management.Automation.PSCredential]
  * [System.Management.Automation.Credential()]
* Optional
* Used to handle 'basic' Authentication
* Exclusive compared to -Certificate

### -Certificate

* Type
  * [X509Certificate]
* Optional
* Used to handle 'Certificate' Authentication
* Exclusive compared to -Credential

### -Session

* Type
  * [AtlassianPS.ServerData]
* Optional
* ParameterSet
  * 'Session'

### -SessionByServerName

* Type
  * [String]
* Replaces and performs the same function as -ServerName. This looks up a Session by Server Name.
* Optional
* ParameterSet
  * 'ServerName'

### -ApiUri

* Type
  * [Uri]
* Path to rest endpoint. Includes both server name and api path. This points to the REST API path
  to handle situations with and without context. See documentation and example below. This would also handle Jira with its API version. When using the "Session" cmdlets, i.e. New-AtlassianSession, the different parts of the host could be optionally specified.
  
  Example when created by *-Session cmdlets, not completely accurate.

  ```ps
  $Session = New-AtlassianSession -Uri "https://myserver.com" -ServerType Confluence
  $Session.ApiUri
  
  #'rest/api' gets added by default. 
  https://myserver.com/rest/api
  
  
  $Session = New-AtlassianSession -Uri "https://myserver.com" -ApiPath "confluence/rest/api" -ServerType Confluence
  $Session.ApiUri
  
  https://myserver.com/confluence/rest/api
  ```

  * Confluence - https://docs.atlassian.com/ConfluenceServer/rest/6.14.1/
    * With context: http://host:port/context/rest/api/resource-name
    * Or without context: http://host:port/rest/api/resource-name
    * Example with context: http://example.com:8080/confluence/rest/api
    * Example without context: http://confluence.myhost.com:8095/rest/api
  * Jira - https://developer.atlassian.com/server/jira/platform/rest-apis/
    * http://host:port/context/rest/api-name/api-version/resource-name
    * https://jira.atlassian.com/rest/api/latest/issue/JRA-9

## Helper Cmdlets

### Parse-AtlassianApiUriFromCommonParameter

* Used to retrieve the ApiUri from all the common parameters
* Example

  ```ps
  $resourceApi = "$(Parse-AtlassianApiUriFromCommonParameter -InputObject $PSBoundParameters)/content/{0}/label?name={1}"
  ```

### Copy-CommonParameters

## Session Cmdlets

### New-AtlassianSession

### Remove-AtlassianSession

### Get-AtlassianSession

### Set-AtlassianSession

### Export-AtlassianSession

### Import-AtlassianSession

## Other Thoughts

### Combine Modules into single repository

Leverage the 'AtlassianPS/AtlassianPS' repository to house all the AtlassianPS modules, i.e. ConfluencePS, JiraPS, AtlassianPS.Configuration,BitbucketPS. The benefit would enable easier development when changes need to occur in multiple modules. It would remove the need to copy the same code, like Invoke-Method.ps1, to multiple repositories with the potential for the share code to get out of sync. It would also allow changes that affect all three modules or a change to AtlassianPS.Configuration that requires changes to the other modules to be handled in on Pull Request.

This single repository could produce multiple modules as is being done today or a single AtlassianPS module with all the current functionality. The extensive changes being made in the v3 branches would make a perfect time for this switch.

**Comment by lipkau** [comment]
Sorry, no time to ready all of that now.
I will edit this comment once I did.

But here you can find my ongoing work on this matter:
https://github.com/lipkau/ConfluencePS/tree/feature/useAtlassianPSConfiguration

**Comment by lipkau** [comment]
First of: thank you for that!

> The ideas for this were based on the patterns found in the *-PSSession cmdlets and the WebRequest/-RestMethod Cmdlets.

Yes. Good cmdlets to base on. I also use the way how [Azure Powershell](https://docs.microsoft.com/en-us/powershell/azure/get-started-azureps?view=azps-1.5.0) connects to the server as an inspiration.

> ## Common Parameters
> [...]
> ### -Credential

✔️ 

> ### -Certificate

I don't see this as something that is used frequently. The modules are over 3 years old and only now code was added for supporting authentication with certificates.
Further, this adds a lot of complexity to the parameterSet: as an anonymous call should be supported, both `-Certificate` and `-Credential` must be optional. and by placing them in different parameterSets, you make the tab completion of the parameters more difficult.
My approach was: do not allow `-Certificate` as a parameter - only usable with Sessions (in the config entry of the server). And if the users complain that it should be a parameter on every cmdlet, add it later (as it would not be a breaking change)

> ### -Session

Why would you want to expose the parameter?
The current implementation of `AtlassianPS.Configuration` (as described in this issue), is based on the premise that the server must be identified for each call (unless `Connect-Server` is used). Therefore the Session is stored for the server and can always be found without having to provide it as a parameter.

What is your argument against the current implementation with `-ServerName`?

> ### -SessionByServerName

same as above

> ### -ApiUri

same as above: What is your argument against the current implementation with `-ServerName`?
Why do you see the need to be able to specify the baseurl for every call of a cmdlet?

I am not quite sure what you meant with _context_ in your examples.
(apart from them not being supported by default because they are not TLS secure :-P)
Is it just that they have a subdomain or a segment for local path?

either way, it is wanted to have the behavior as shown here, where the cmdlet defines the endpoint with the same string as it is in the api documentation (not appending `rest/api` to the ApiUri):
https://github.com/lipkau/ConfluencePS/blob/a8a5260ce7d74788f2c6f779dc72176f0b633a02/ConfluencePS/Public/Get-Space.ps1#L38

> ## Helper Cmdlets
> ### Parse-AtlassianApiUriFromCommonParameter

My comments from above make this unnecessary, as the `ApiUri` is the baseurl in the ServerData.

> ## Session Cmdlets
> ### New-AtlassianSession

Excellent idea, to make this a shared command.

> ### Remove-AtlassianSession

Touché. This command is indeed missing.

> ### Get-AtlassianSession

Is missing. But I don't see a high demand for this.

> ### Set-AtlassianSession

Is missing. But I don't see a high demand for this.

> ### Export-AtlassianSession
> ### Import-AtlassianSession

Are both not wanted, as they present a security risk.
_An experienced powershell user would be able to do this with `Export-Clixml`._

> ## Other Thoughts
> ### Combine Modules into single repository

I have to disagree with you there.
The time and effort people have to put into the modules is very different. The demand for the modules also differs a lot.

If all AtlassianPS modules were in a single repository, the maintenance and publishing would be considerably more difficult.

However!!!
> It would remove the need to copy the same code, like Invoke-Method.ps1, to multiple repositories with the potential for the share code to get out of sync.

This is something I thought of a long time ago. To have a module which only includes stuff common to all modules. `Invoke-Method` is a very good example. Because at first glance, it appears as it can be copy pasted between modules - but some small things are different (api arguments for paged results).
With `AtlassianPS.Configuration` I am already aiming at making these differences part of a configuration so that it would be possible to use the exact same, generic, `Invoke-Method` in both modules.

--------------------------------

Here is my ongoing work on v3 for ConfluencePS:
https://github.com/lipkau/ConfluencePS/tree/feature/useAtlassianPSConfiguration

Here is my most recent change for `AtlassianPS.Configuration`, which was to support Certificates:
https://github.com/AtlassianPS/AtlassianPS.Configuration/commit/555e8a3300a12e03b1d9c494a874fe21fe9bbccf

**Comment by ritzcrackr** [comment]
Thanks for the consideration and feedback. The main reason for having these parameters on every function is to support the single function scenario that I describe below. The names I picked for the parameters most accurately described the purpose of the parameter IMHO. 

As I have said other place, I just want to ensure that any code I contribute will be accepted and I don't waste mine or your time.

## Naming and Purpose of suggested Parameters 

> -Session
> 
> [...]
> 
> -SessionByServerName
> 
> [...]
>
> -ApiUri
> 
> [...]
> 

I completely understand that naming is hard. The reason I am using `Session` is that I believe it more accurately describes the output of `Connect-Server`. I am equating the output of `Connect-Sever` to `New-PSSession` or the combination of `-WebSession` and `-SessionVariable` on `Invoke-WebRequest`. I equate `-ServerName` to `-ComputerName` on `New-PSSession` or `Enter/Connect-PSSession` and not to the concept of a persistent state. 

The reason for both `-Session` and `-SessionByServerName` was to support your existing concept of being able to lookup an existing session by a `ServerName` using `-SessionByServerName` and passing an existing session using `-Session`.

```powershell
$Session = New-AtlassianSession -ServerName "ConfluenceServer" [...]

Get-Page -Session $Session [...]
```

versus

```powershell
New-AtlassianSession -ServerName "ConfluenceServer" [...]

Get-Page -SessionByServerName "ConfluenceServer" [...]
```

The need to specify a combination of `-Credential`, `-Certification`, `-Session`, `-SessionByServerName`, or `ApiUri` is to seamlessly handle multiple use cases of these cmdlets. Once again I am basing it on `Enter/Connect-PSSession` versus `New-PSSession`. There are situations where you are just making a single call and there is no need to create a session. I don't believe that a call to `Connect-Server` should be necessary when you just want to call a single function. This is currently support by the modules today but if `-ApiUri` is remove or not present on all functions then this ability goes away.

```powershell
Get-Page -ApiUri "https://confluence.com/rest/api/" -Certificate [...]
```

versus a script with multiple calls to an Atlassian server.

```powershell
$Session = New-AtlassianSession -ServerName "ConfluenceServer" -ApiUri "https://confluence.com/rest/api/" -Certificate [...]

Get-Page -Session $Session [...]
Add-Page -Session $Session [...]
#...more stuff
Get-AddLabel -Session $Session [...]
```

## Certificate Parameter
>> -Certificate

> I don't see this as something that is used frequently. The modules are over 3 years old and only now code was added for supporting authentication with certificates.
> Further, this adds a lot of complexity to the parameterSet: as an anonymous call should be supported, both -Certificate and -Credential must be optional. and by placing them in different parameterSets, you make the tab completion of the parameters more difficult.
> My approach was: do not allow -Certificate as a parameter - only usable with Sessions (in the config entry of the server). And if the users complain that it should be a parameter on every cmdlet, add it later (as it would not be a breaking change)

The reason to add `-Certificate` now is several reasons. First, if I update all of the parameters I might as well add all of them. Second, I need certificate support and if I am doing the work then I need to be able to use the output. I am the complaining user :smile:  Thirdly, as you stated adding this would not be a breaking change. If a user doesn't need the parameter then it wouldn't affect their existing scripts. 

`-certificate` could be lower down in the parameter order which would make tab completion easier. I put authentication parameters in different parameter sets since they should be used exclusive to each other.

## -ApiUri and 'context'

>I am not quite sure what you meant with context in your examples. 
>
>(apart from them not being supported by default because they are not TLS secure :-P)
> 
> Is it just that they have a subdomain or a segment for local path?

The 'context' is an Atlassian concept. I pulled those links about 'context' and Uri format from the Atlassian Confluence and Jira API documentation. Reviewing the AtlassianPS.Configuration documentation again it looks like the ability to specify the 'context' or the full URI already exists. I think I was confused by how each cmdlet appended `/rest/api/resource-name` to the Uri. A reason to have the ability to specify the `/rest/api` part of the Uri is that at least with the Jira rest api there is the ability to specify different versions of rest API. `/rest/api/latest/issue` versus `/rest/api/6.0/issue`. Not sure that the ability to specify the Rest API version is a great reason for additional complexity but it could be useful.

## Session Cmdlets

:heavy_check_mark:

## Other thoughts

I am fine with your reasons. Just wanted to make the suggestions since I am a big fan of the [monorepo](https://en.wikipedia.org/wiki/Monorepo) concept. In my case I would have added the Certificate parameter to both ConfluencePS and JiraPS if all the code was in one repo but since I needed to make two PRs I didn't. I view these three repositories as one project and I would combine them but you are running this project and I was just making a suggestion.

[Azure Powershell repository](https://github.com/Azure/azure-powershell/tree/master/src) layout is an example of how I would combine all the AtlassianPS modules under a single repository.

**Comment by KevinMarquette** [comment]
A pattern that I like for modules like this is to leverage `Sessions` that contain all the connectivity related details. So have the full set of commands for session management with support for setting the active or default session. Then every command gets an optional `-Session` parameter. If that is not provided, the command will call the Get for the active/default session. 

All those other parameters like URI, Credential, Certificate, ect would then be part of the session and not on the individual commands. This removes a lot of copy/paste structure from all those commands and makes them easier to maintain. It also frees up your parameter sets to be better used by your functions.

I often take it to the extreme and make the session a class that would expose a method like your `Invoke-Method`. This lets the functions just call that method on the class and the class would internally manage the use of all those parameters. But PowerShell classes can bring their own issues and you could just as easily pass the session to a `Invoke-Method`.

On a side note, the calls to `New-Session` should try to perform some lightweight request so if there are authentication\connection issues, they are discovered and error on that call instead on the other functions.

**Comment by lipkau** [comment]
_sorry for the *very* late response_

@ritzcrackr
JiraPS currently doesn't support a _single function scenario_ as you described.
But ConfluencePS does.

I discussed this topic extensively with several users and powershell MVPs.
I would prefer not to make the changes you proposed *for now*.
We can create an issue in AtlassianPS/AtlassianPS describing your proposal and leave it for voting (with reactions).

In case you are still interested in helping out with a PR, I would like you to stick to the example in my branch for ConfluencePS.
